components/clarinet-format/src/formatter/mod.rs

```diff
@@ -1981,4 +1989,13 @@ mod tests_formatter {
         let result = format_with_default(src);
         assert_eq!(expected, result);
     }
+
+    #[test]
+    fn format_ast_without_source() {
+        let src = "(define-private (noop) (ok true))";
+        let ast = clarity::vm::ast::parser::v2::parse(src).unwrap();
+        let formatter = ClarityFormatter::new(Settings::default());
+        let result = formatter.format_ast(&ast);
+        assert_eq!(result, "(define-private (noop)\n  (ok true)\n)\n");
+    }
```

Comment:
Could you add this as well? I don't think `ignored_exprs` will work without the source string
```
    #[test]
    fn format_ast_without_source_ignore() {
        let src = ";; @format-ignore\n(define-private (noop) (ok true))";
        let ast = clarity::vm::ast::parser::v2::parse(src).unwrap();
        let formatter = ClarityFormatter::new(Settings::default());
        let result = formatter.format_ast(&ast);
        assert_eq!(result, src);
    }
```

Reply:
Indeed, that is precisely the kind of edge case I wouldn't have anticipated.
But that would be the expected behavior, right?

If we only have an AST and no source code as a "source of truth", it wouldn't make sense to "ignore formatting". The idea is to be able to take a raw AST, possibly with no span at all, and format it in text.

I think it makes sense to me

Reply:
That's what I am currently exploring with the transpiler where I go from TS AST to Clarity AST/PSE (and then AST to clarity source code with the formatter). So there's no Clarity source code